Remove the jetpack entrypoint - take 3.

enejb opened the pull request:
This PR is based on https://github.com/Automattic/wp-calypso/pull/39708

#### Changes proposed in this Pull Request
Remove the current jetpack cloud entry point. 

<img width="1116" alt="Screen Shot 2020-02-27 at 4 00 08 PM" src="https://user-images.githubusercontent.com/115071/75455991-4105c200-597a-11ea-8cd6-618027122a30.png">


Some regression in this PR. 
1. The design falls back to calypso. (no more blackmasterbar) since we are not including the styles anymore.
2. Doesn't limit the sites in jetpack cloud  to just jetpack sites. 

I think they should be addressed in a different PR.

#### Testing instructions
1. Start up calypso `npm start` 
2. notice that nothing seems to be broken and that we can navigate to all the sections as expected. 
3. Notice that you can't visit sections like `scan` and `backup`

Test cloud. 
1. Start up jetpack cloud `npm start start-jetpack-cloud` 
2. Notice the regressions but you should be able to visit all the other section as expected.

enejb committed:
Remove the jetpack entrypoint - take 3.

This version of the PR doesn't make any changes to main jetpack-cloud controller

enejb committed:
Revert Calypso masterbar and sidebar.

jsnajdr commented on client/server/bundler/sections-loader.js:
```diff
@@ -52,7 +60,11 @@ const loader = function() {
+	// Build all the sections during the build step.
+	let sections = forceRequire
+		? require( this.resourcePath )
+		: require( this.resourcePath ).filter( utils.filterSections );
```
This section-filtering code reads from `config`, depends on `CALYPSO_ENV`, and runs at build time, as part of a webpack plugin/loader.

I think it would be better to postpone the section filtering to runtime. I.e., generate only one production bundle that can run as Calypso (in any environment like `stage` or `horizon`) or Jetpack Cloud, depending on how it's run:
```
CALYPSO_ENV=jetpack-cloud-stage node build/bundle.js
```

Is there an agreement on this? Or are there important reasons why there should be two bundles generated?

I can suggest ways to do the section filtering at runtime if we agree on the premise.

enejb replied:
There are a couple webpack configs. This uses all the sections when the `webpack.config.node.js` is run and it uses sections from the config when `webpack.config.js` is being used.  

Another place where we are doing the section filtering is on pages/index.js see https://github.com/Automattic/wp-calypso/pull/40022/files#diff-13202bbd2bfb14a1570c13babd9e8444R825 

Is this what you mean by doing filtering at runtime?

jsnajdr replied:
The `webpack.config.node.js` config is used to build the server that runs with Node.js. This Express.js-based HTTP server does two things related to sections:
- for routes like `/stats/*`, makes sure that the JS chunks that are required to load Stats are included as `<script>` tags in the HTTP response. You can see this in action when opening `https://wordpress.com/stats/day` in browser and inspecting the HTML document you got in response:

<img width="607" alt="Screenshot 2020-03-11 at 11 31 36" src="https://user-images.githubusercontent.com/664258/76407658-fded1b00-638b-11ea-8189-10b15a66a353.png">

There's the `entry-main` chunk, there's `stats` and also all the shared chunks. The server knows what to send because each section has its own server-side handler.

- for isomorphic sections like `themes` or `log-in`, the server renders the complete HTML markup of the page.

`webpack.config.js` is used to build the JS bundle that runs in the browser. At this moment, because the section filtering is done at build-time, separate builds are required for Calypso and Jetpack Cloud.

Ideally, we could run just one Docker image for all and just set the `CALYPSO_ENV` environment variable to choose which app and environment exactly we want to run.

jsnajdr replied:
In https://github.com/Automattic/wp-calypso/pull/40022/commits/5649277c2e05c528e8800e41c3e8df2706b2f31d I implemented the runtime filtering: `section-loader` will build all sections (it adds some generated code to the `sections.js` module and leaves it intact otherwise), and the code (both server and client) will the filter it at runtime according to the currently active `CALYPSO_ENV`.

enejb committed:
Remove more sidebar changes

jsnajdr committed:
Do section filtering always at runtime

Remove section filtering from `section-loader` that is performed at build time.

Create a new module called `sections-filter`. It contains code that will run in both Node.js
and browser. `server/bundler/utils` has Node-specific code that depends on modules like `crypto`
or `fs`.

Use the `isSectionEnabled` function to filter the sections list at runtime, both on server and client.
It will filter according to the `config` that is active at runtime.

jsnajdr committed:
Fix module name typo

jsnajdr committed:
sections-loader: remove unnecessarily verbose logging

jsnajdr reviewed (approved):
Ready to :ship:
